### PR TITLE
[HttpClient] improve doc again

### LINKED ISSUE #11530: Complete the HttpClient documentation
HttpClient is a new Symfony 4.3 component. We have some docs for it: https://symfony.com/doc/master/components/http_client.html  but the latest features added to it may be missing.

### PR DESCRIPTION
**Pull request by nicolas-grekas**
Fixes all todos and more :)

fixes #11530.

### CODE REVIEW

File: components/http_client.rst [hunk 1]
```diff
@@ -317,6 +321,146 @@ When the HTTP status code of the response is in the 300-599 range (i.e. 3xx,
...
+
+    Use the ``user_data`` option combined with ``$response->getInfo('user_data')``
+    to track the identity of the responses in your foreach loops.
+
+Dealing with network timeouts
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This component allows dealing with both request and response timeouts.
+
+A timeout can happen when e.g. DNS resolution takes to much time, when the TCP
```
Comment (javiereguiluz) [inline components/http_client.rst]: `to much time` -> `too much time`

File: components/http_client.rst [hunk 2]
```diff
@@ -317,6 +321,146 @@ When the HTTP status code of the response is in the 300-599 range (i.e. 3xx,
...
+connection cannot be opened in the given time budget, or when the response
+content pauses for too long. This can be configured by the ``timeout`` request
```
Comment (javiereguiluz) [inline components/http_client.rst]: `configured by` -> `configured with` ?

File: components/http_client.rst [hunk 3]
```diff
@@ -317,6 +321,146 @@ When the HTTP status code of the response is in the 300-599 range (i.e. 3xx,
...
+    foreach ($client->stream($responses, 1.5) as $response => $chunk) {
+        if ($chunk->isTimeout()) {
+            // $response staled for more than 1.5 seconds
+        }
+    }
+
+A timeout is not necessarily an error: you can decide to stream again the
+response and get remaining contents that might come back in a new timeout, etc.
+
+.. tip:
```
Comment (javiereguiluz) [inline components/http_client.rst]: Minor RST syntax issue. The `tip` directive must end with two colons.

File: components/http_client.rst [hunk 4]
```diff
@@ -317,6 +321,146 @@ When the HTTP status code of the response is in the 300-599 range (i.e. 3xx,
...
+
+    Passing ``0`` as timeout allows monitoring responses in a non-blocking way.
+
+.. note:
```
Comment (javiereguiluz) [inline components/http_client.rst]: Same here with the `note` directive.

**Review by javiereguiluz — APPROVED** [review]
Very nice 👏  I left some minor comments. Thanks!

File: components/http_client.rst (hunk 1 above)
Comment (nicolas-grekas) [reply components/http_client.rst]: fixed thanks

File: components/http_client.rst [hunk 5]
```diff
@@ -317,6 +321,146 @@ When the HTTP status code of the response is in the 300-599 range (i.e. 3xx,
...
+            // of the response body that just arrived
+        }
+    }
+
+.. tip::
+
+    Use the ``user_data`` option combined with ``$response->getInfo('user_data')``
+    to track the identity of the responses in your foreach loops.
+
+Dealing with network timeouts
```
Comment (wouterj) [inline components/http_client.rst]:
```suggestion
Dealing with Network Timeouts
```

(title case)

File: components/http_client.rst [hunk 6]
```diff
@@ -317,6 +321,146 @@ When the HTTP status code of the response is in the 300-599 range (i.e. 3xx,
...
+
+    Passing ``0`` as timeout allows monitoring responses in a non-blocking way.
+
+.. note::
+
+    Timeouts control how long one is willing to wait *while the HTTP transation
+    is idle*. Big responses can last as long as needed to complete, provided they
+    remain active during the transfer and never pause for longer than specified.
+
+Dealing with network errors
```
Comment (wouterj) [inline components/http_client.rst]:
```suggestion
Dealing with Network Errors
```

File: components/http_client.rst [hunk 7]
```diff
@@ -317,6 +321,146 @@ When the HTTP status code of the response is in the 300-599 range (i.e. 3xx,
...
+
+    Because ``$response->getInfo()`` is non-blocking, it shouldn't throw by design.
+
+When multiplexing responses, you can deal with errors for individual streams by
+catching ``TransportExceptionInterface`` in the foreach loop::
+
+    foreach ($client->stream($responses) as $response => $chunk) {
+        try {
+            if ($chunk->isLast()) {
+                // do something with $response
```
Comment (wouterj) [inline components/http_client.rst]:
```suggestion
                // ... do something with $response
```

**Review by wouterj — APPROVED** [review]

File: components/http_client.rst (hunk 5 above)
Comment (nicolas-grekas) [reply components/http_client.rst]: fixed thanks

File: components/http_client.rst [hunk 8]
```diff
@@ -317,6 +321,146 @@ When the HTTP status code of the response is in the 300-599 range (i.e. 3xx,
...
+.. note::
+
+    Timeouts control how long one is willing to wait *while the HTTP transation
+    is idle*. Big responses can last as long as needed to complete, provided they
+    remain active during the transfer and never pause for longer than specified.
+
+Dealing with Network Errors
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Network errors (broken pipe, failed DSN resolution, etc.) are thrown as instances
```
Comment (andreia) [inline components/http_client.rst]:
```suggestion
Network errors (broken pipe, failed DNS resolution, etc.) are thrown as instances
```

File: components/http_client.rst [hunk 9]
```diff
@@ -233,13 +232,12 @@ making a request. Use the ``max_redirects`` setting to configure this behavior
...
-..
-.. TODO
-..
-..
+Advanced Options
+~~~~~~~~~~~~~~~~
+
+The :class:`Symfony\\Contracts\\HttpClient\\HttpClientInterface` defines all the
+options you might need to take full control of the way the request is performed,
+including progress monitoring, DSN pre-resolution, timeout, SSL parameters, etc.
```
Comment (andreia) [inline components/http_client.rst]:
```suggestion
including progress monitoring, DNS pre-resolution, timeout, SSL parameters, etc.
```

File: components/http_client.rst [hunk 10]
```diff
@@ -317,6 +321,146 @@ When the HTTP status code of the response is in the 300-599 range (i.e. 3xx,
...
+A timeout is not necessarily an error: you can decide to stream again the
+response and get remaining contents that might come back in a new timeout, etc.
+
+.. tip::
+
+    Passing ``0`` as timeout allows monitoring responses in a non-blocking way.
+
+.. note::
+
+    Timeouts control how long one is willing to wait *while the HTTP transation
```
Comment (andreia) [inline components/http_client.rst]:
```suggestion
    Timeouts control how long one is willing to wait *while the HTTP transaction
```

File: components/http_client.rst [hunk 11]
```diff
@@ -317,6 +321,146 @@ When the HTTP status code of the response is in the 300-599 range (i.e. 3xx,
...
+
+Multiplexing Responses
+~~~~~~~~~~~~~~~~~~~~~~
+
+If you look again at the snippet above, responses are read in requests' order.
+But maybe the 2nd response came back before the 1st? Fully asynchronous operations
+require being able to deal with the responses in whatever order they come back.
+
+In order to do so, the ``stream()`` method of HTTP clients accepts a list of
+responses to monitor. As mentionned :ref:`previously <http-client-streaming-responses>`,
```
Comment (andreia) [inline components/http_client.rst]:
```suggestion
responses to monitor. As mentioned :ref:`previously <http-client-streaming-responses>`,
```

File: components/http_client.rst [hunk 12]
```diff
@@ -317,6 +321,146 @@ When the HTTP status code of the response is in the 300-599 range (i.e. 3xx,
...
+content pauses for too long. This can be configured with the ``timeout`` request
+option::
+
+    // A TransportExceptionInterface will be issued if nothing
+    // happens for 2.5 seconds when accessing from the $response
+    $response = $client->request('GET', 'https://...', ['timeout' => 2.5]);
+
+The ``default_socket_timeout`` PHP ini setting is used if the option is not set.
+
+The option can be overriden by using the 2nd argument of the ``stream()`` method.
```
Comment (andreia) [inline components/http_client.rst]:
```suggestion
The option can be overridden by using the 2nd argument of the ``stream()`` method.
```

File: components/http_client.rst [hunk 13]
```diff
@@ -476,18 +621,11 @@ If you have several clients, you must use any of the methods defined by Symfony
 to ref:`choose a specific service <services-wire-specific-service>`. Each client
```
Comment (andreia) [inline components/http_client.rst]:
```suggestion
to :ref:`choose a specific service <services-wire-specific-service>`. Each client
```

File: components/http_client.rst [hunk 14]
```diff
@@ -265,6 +263,12 @@ following methods::
     // you can get individual info too
     $startTime = $response->getInfo('start_time');
 
+.. note::
+
+    ``$response->getInfo()`` is non-blocking: it returns *live* information
+    about the response. Some of them might not be know yet (e.g. ``http_code``)
```
Comment (andreia) [inline components/http_client.rst]:
```suggestion
    about the response. Some of them might not be known yet (e.g. ``http_code``)
```

**Review by andreia — CHANGES_REQUESTED** [review]
Thanks a lot for documenting this :)
( I just made some tweaks to the cheat sheet also based on your update, thank you! :) )

**Comment by javiereguiluz** [comment]
Nicolas, thanks a lot for contributing these improvements. Thanks to reviewers too, specially Andréia ... I fixed all the typos you reported while merging. Thanks!

**Comment by andreia** [comment]
Thank you Javier! :)

**Pull request merged**